src/main/kotlin/org/jetbrains/kotlinx/jupyter/repl/impl/InternalEvaluatorImpl.kt

```diff
@@ -124,4 +158,73 @@ internal class InternalEvaluatorImpl(
             isExecuting = false
         }
     }
+
+    private fun updateVariablesState(cellId: Int) {
+        // remove known modifying usages in this cell
+        usageMap[cellId]?.removeIf {
+            containingMap[it] != cellId
+        }
+
+        variablesMap.forEach {
+            val state = it.value
+            val property = state.propertyKlass
+            val newValue = property.get(state.scriptInstance) ?: return@forEach
+
+            if (state.stringValue != newValue.toString()) {
+                usageMap[cellId]?.add(it.key)
+                state.stringValue = newValue.toString()
+            }
+        }
+    }
+
+    private fun getVisibleVariables(target: ResultValue, cellId: Int): Map<String, VariableState> {
+        val klass = target.scriptClass ?: return emptyMap()
+        val cellKlassInstance = target.scriptInstance!!
+
+        val fields = klass.declaredMemberProperties
+        val ans = mutableMapOf<String, VariableState>()
+        fields.forEach {
+            val casted = it as KProperty1<Any, *>
+            ans[casted.name] = if (casted.get(cellKlassInstance) == null) {
+                return@forEach
+            } else {
+                val stringVal = casted.get(cellKlassInstance).toString()
+                val state = VariableState(casted, cellKlassInstance, stringVal)
+
+                // redeclaration of any type
+                if (containingMap.containsKey(casted.name)) {
+                    val oldCellId = containingMap[casted.name]
+                    if (oldCellId != cellId) {
+                        usageMap[oldCellId]?.remove(casted.name)
+                    }
+                }
+                // it was val, now it's var
+                if (it is KMutableProperty1 && variablesCache.containsKey(casted.name)) {
+                    variablesCache.remove(casted.name)
```

Comment:
What if we call `remove` on absent key? I think, this condition may be simplified